File: backend/app/schemas/listing.py
from datetime import date
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel


class ListingResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    macro_location: str
    city: str
    pin_code: str
    property_type: str
    status: str
    report_count: int
    vendor_count: int
    latest_report_date: date | None = None


class ListingBrowseResponse(BaseModel):
    listings: list[ListingResponse]
    total: int
    page: int
    page_size: int


class RedactedReportPreview(BaseModel):
    id: UUID
    report_category: str
    locality: str | None = None
    city: str | None = None
    pin_code: str | None = None
    property_type: str | None = None
    plot_extent_sqft: int | None = None
    built_up_sqft: int | None = None
    report_date: date | None = None
    latitude: float | None = None
    longitude: float | None = None
    content_preview: dict | None = None
    is_purchased: bool = False


class ListingDetailResponse(BaseModel):
    listing: ListingResponse
    reports: list[RedactedReportPreview]


class VendorListingReportItem(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    report_category: str
    property_address: str | None = None
    city: str | None = None
    pin_code: str | None = None
    property_type: str | None = None
    report_date: date | None = None
    status: str
    listing_approved: bool


class VendorListingGroup(BaseModel):
    listing: ListingResponse
    reports: list[VendorListingReportItem]


class VendorListingsResponse(BaseModel):
    groups: list[VendorListingGroup]
    total: int
    page: int
    page_size: int

File: frontend/src/types/listing.ts
import { Report } from "./report";

export interface ListingResponse {
  id: string;
  macro_location: string;
  city: string;
  pin_code: string;
  property_type: string;
  status: string;
  report_count: number;
  vendor_count: number;
  latest_report_date: string | null;
}

export interface ListingBrowseResponse {
  listings: ListingResponse[];
  total: number;
  page: number;
  page_size: number;
}

export interface RedactedReportPreview {
  id: string;
  report_category: string;
  locality: string | null;
  city: string | null;
  pin_code: string | null;
  property_type: string | null;
  plot_extent_sqft: number | null;
  built_up_sqft: number | null;
  report_date: string | null;
  latitude: number | null;
  longitude: number | null;
  content_preview: Record<string, string | number | null> | null;
  is_purchased: boolean;
}

export interface ListingDetailResponse {
  listing: ListingResponse;
  reports: RedactedReportPreview[];
}

export interface VendorListingReportItem {
  id: string;
  report_category: string;
  property_address: string | null;
  city: string | null;
  pin_code: string | null;
  property_type: string | null;
  report_date: string | null;
  status: string;
  listing_approved: boolean;
}

export interface VendorListingGroup {
  listing: ListingResponse;
  reports: VendorListingReportItem[];
}

export interface VendorListingsResponse {
  groups: VendorListingGroup[];
  total: number;
  page: number;
  page_size: number;
}

export interface PurchaseResponse {
  id: string;
  report_id: string;
  listing_id: string;
  lender_id: string;
  price: string;
  created_at: string;
}

export interface PurchasedReportItem {
  purchase: PurchaseResponse;
  report: Report;
}

export interface PurchasedReportsResponse {
  items: PurchasedReportItem[];
  total: number;
  page: number;
  page_size: number;
}
